### ISSUE TITLE
vm run-command --script not working on Linux VMs

### ISSUE THREAD

**Issue by thisSIDEofRANDOM** [Azure/azure-cli#5398]
When trying to use run-command with a file as the argument against a linux target it is consistently failing. On inspection of the file some odd escape characters are being inserted when it copies the file on to the VM for execution:

The documentation mentions:
```
--scripts
Script lines separated by whites spaces. Use @{file} to load from a file.
```

I can't seem to figure out what its wanting me to pass this file as to make it run correctly. Its uploading to the server but inserting the odd escape characters as seen below. Didn't seen any documentation on what to do here.

### Local file:
```
# cat -v test.sh 
logger "test"
```
### Command Executed to run against vm and failure message:
```
bash-4.3# az vm run-command invoke --resource-group 'TEST' --name 'TEST1' --command-id 'RunShellScript' --scripts @./test.sh --verbose
Deployment failed. Correlation ID: -------. VM has reported a failure when processing extension 'RunCommandLinux'. Error message: "Enable failed: failed to execute command: command terminated with exit status=126
[stdout]

[stderr]
/bin/sh: /var/lib/waagent/run-command/download/6/script.sh: cannot execute binary file
".
```

### File on the server with odd characters:
```
# cat -v /var/lib/waagent/run-command/download/6/script.sh
l^@o^@g^@g^@e^@r^@ ^@"^@t^@e^@s^@t^@"^@
```
Tried running this with multiple iterations, single lines, multiple. Declaratives.

I thought for a bit @{} was referencing a hashmap I.E. powershell but trying different variations of loading the file in a array didn't produce any tangible results.

It almost looks like it is compressing the file on the way up and incorrectly uncompressing it.

---

### Environment summary

Install Method Docker / Arch / Bash 

When trying to run az vm run-command invoke --scripts the text in our script file seems to be getting converted and having odd characters inserted?

**Comment by thisSIDEofRANDOM** [comment]
Oddly seems to work when I gzip the scripts up before sending them to the server in some cases, not all.

**Comment by thisSIDEofRANDOM** [comment]
Peaking at the file post upload failure and it looks like its converting to double bye on upload with the second byte as null?

BYTE HEX ASCII output 

00000000  6C 00 6F 00 67 00 67 00 65 00 72 00 20 00 22 00 74 00 65 00 73 00 74 00 l.o.g.g.e.r. .".t.e.s.t.
00000018  22 00

**Comment by yugangw-msft** [comment]
@thisSIDEofRANDOM, maybe just a typo when you reported the issue, but the command has a white-space between `@` and `./test.sh`. Once I removed it,  the command worked. FYI, the `@` is a CLI specific pre-processor like mechanism, that the parser will try to read the file following the `@` and feed the whole content to the command.

**Comment by yugangw-msft** [comment]
Also, the `run-command` is a thin wrapper around the custom scripts extension, nothing more really. So you can use `az vm extension set` to cross check.

**Comment by yugangw-msft** [comment]
please re-activate if you are still blocked

**Comment by thisSIDEofRANDOM** [comment]
@yugangw-msft please reactivate. This is still happening and yes it was a typo.

Can you confirm if you were running from a Linux box to a Linux box. This happens on every server we try against coming from Ubuntu,Centos,Debian,Arch,Docker, and also windows subsystem against Linux machines.

The script which is written in VI uploads to the machine and then fails to execute as every character looks to have been converted to a double byte with the first part NULL meaning it wont run.

If you can share what your script looked like as well as what verison of azure-cli you were using and what type of machine you were running on that might help narrow it down but as of right now we see this issue consistently and can reproduce it.

**Comment by yugangw-msft** [comment]
I ran from OSX/Terminal to UbuntuL16.04 VM, and I used vim and input the same dummy command `logger "test"`, then I invoked the same command like yours

**Comment by yugangw-msft** [comment]
I can try from ubuntu to ubuntu

**Comment by thisSIDEofRANDOM** [comment]
Thanks @yugangw-msft we don't have a OSX machine handy to test on. or I would give that a shot. Let me know if I can grab any other information on the machines we have tested with. I'm at a pretty heavy loss on the behavior.

**Comment by yugangw-msft** [comment]
Works on linux client for me
```bash
(env) yugangw@yg-linux:~/azure-cli$ az vm run-command invoke --resource-group yg2 --name yg2-1 --command-id 'RunShellScript' --scripts @./test.sh
{
  "endTime": "2018-01-26T23:29:27.908795+00:00",
  "error": null,
  "name": "706c33d0-173e-4f62-9083-18bf387d39a5",
  "output": [
    {
      "code": "ProvisioningState/succeeded",
      "displayStatus": "Provisioning succeeded",
      "level": "Info",
      "message": "Command is finished.\n---stdout---\n\n---errout---\n\n"
    }
  ],
  "startTime": "2018-01-26T23:28:58.971315+00:00",
  "status": "Succeeded"
}
(env) yugangw@yg-linux:~/azure-cli$ az vm run-command invoke --resource-group yg2 --name yg2-1 --command-id 'RunShellScript' --scripts 'echo hello'{
  "endTime": "2018-01-26T23:30:04.908797+00:00",
  "error": null,
  "name": "134cfbfd-a18a-49d3-b550-4034b2ff52b0",
  "output": [
    {
      "code": "ProvisioningState/succeeded",
      "displayStatus": "Provisioning succeeded",
      "level": "Info",
      "message": "Command is finished.\n---stdout---\nhello\n\n---errout---\n\n"
    }
  ],
  "startTime": "2018-01-26T23:29:53.783807+00:00",
  "status": "Succeeded"
}

```

**Comment by thisSIDEofRANDOM** [comment]
Hmmm,

Thanks for trying again @yugangw-msft let me try again on our end with a fresh built VM. Maybe someone inserted something odd in our builds.

**Comment by thisSIDEofRANDOM** [comment]
Hi @yugangw-msft brand new RHEL 7.4 server built from marketplace. Tried in latests docker version of Azure CLI and experienced the same problem. Can you confirm you have the same version of az-cli I linked mine below. I copied your echo hello command as well with no luck:

### run-command 
```
bash-4.3# az vm run-command invoke --resource-group tester --name testa --command-id 'RunShellScript' --scripts @./test.sh
Deployment failed. Correlation ID: 05dce6f0-250a-40f4-808c-5f7a8bb1c70a. VM has reported a failure when processing extension 'RunCommandLinux'. Error message: "Enable failed: failed to execute command: command terminated with exit status=126
[stdout]

[stderr]
/bin/sh: /var/lib/waagent/run-command/download/0/script.sh: cannot execute binary file
"
```

### script post upload
```
cat -v script.sh 
e^@c^@h^@o^@ ^@"^@t^@e^@s^@t^@"^@
```

### Tried running with just a script per above
```
bash-4.3# az vm run-command invoke --resource-group tester --name testa --command-id 'RunShellScript' --scripts 'echo hello'
Deployment failed. Correlation ID: f4b2378b-3f2e-4777-bc31-419b15d5abc2. VM has reported a failure when processing extension 'RunCommandLinux'. Error message: "Enable failed: failed to execute command: command terminated with exit status=126
[stdout]

[stderr]
/bin/sh: /var/lib/waagent/run-command/download/2/script.sh: cannot execute binary file
"
```

### az-cli version
```
bash-4.3# az --version
azure-cli (2.0.25)

acr (2.0.19)
acs (2.0.24)
advisor (0.1.1)
appservice (0.1.24)
backup (1.0.6)
batch (3.1.8)
batchai (0.1.4)
billing (0.1.7)
cdn (0.0.11)
cloud (2.0.11)
cognitiveservices (0.1.10)
command-modules-nspkg (2.0.1)
configure (2.0.13)
consumption (0.2.1)
container (0.1.16)
core (2.0.25)
cosmosdb (0.1.16)
dla (0.0.17)
dls (0.0.19)
eventgrid (0.1.8)
extension (0.0.7)
feedback (2.0.8)
find (0.2.8)
interactive (0.3.13)
iot (0.1.16)
keyvault (2.0.16)
lab (0.0.15)
monitor (0.1.0)
network (2.0.21)
nspkg (3.0.1)
profile (2.0.17)
rdbms (0.0.11)
redis (0.2.11)
reservations (0.1.1)
resource (2.0.21)
role (2.0.17)
servicefabric (0.0.9)
sql (2.0.19)
storage (2.0.23)
vm (2.0.24)

Python location '/usr/local/bin/python'
Extensions directory '/root/.azure/cliextensions'

Python (Linux) 3.6.3 (default, Dec 20 2017, 01:32:21) 
[GCC 5.3.0]

Legal docs and information: aka.ms/AzureCliLegal
```

**Comment by thisSIDEofRANDOM** [comment]
@yugangw-msft I tried updating to the most recent version of az-cli as well as same results with commands or scripts.

Are you on a version newer? I can try rolling back if so.

### az version
```
bash-4.3# az --version
azure-cli (2.0.26)

acr (2.0.20)
acs (2.0.25)
advisor (0.1.1)
appservice (0.1.25)
backup (1.0.6)
batch (3.1.9)
batchai (0.1.5)
billing (0.1.7)
cdn (0.0.12)
cloud (2.0.12)
cognitiveservices (0.1.10)
command-modules-nspkg (2.0.1)
configure (2.0.13)
consumption (0.2.1)
container (0.1.16)
core (2.0.26)
cosmosdb (0.1.17)
dla (0.0.18)
dls (0.0.19)
eventgrid (0.1.9)
extension (0.0.8)
feedback (2.0.8)
find (0.2.8)
interactive (0.3.15)
iot (0.1.16)
keyvault (2.0.17)
lab (0.0.16)
monitor (0.1.1)
network (2.0.22)
nspkg (3.0.1)
profile (2.0.18)
rdbms (0.0.11)
redis (0.2.11)
reservations (0.1.1)
resource (2.0.22)
role (2.0.17)
servicefabric (0.0.9)
sql (2.0.20)
storage (2.0.24)
vm (2.0.25)

Python location '/usr/local/bin/python'
Extensions directory '/root/.azure/cliextensions'

Python (Linux) 3.6.3 (default, Dec 20 2017, 01:32:21) 
[GCC 5.3.0]

Legal docs and information: aka.ms/AzureCliLegal
```

**Comment by yugangw-msft** [comment]
I am on the latest, but i don't think the CLI version would matter. 
I tried RHEL which works fine. I suggest 2 things, hope to get a repro
1. Please run `az vm show --resource-group tester --name testa --query 'storageProfile' -o json` and share the output. This tells me the exact image you are using.
2. Use [custom script extension](https://docs.microsoft.com/en-us/azure/virtual-machines/linux/extensions-customscript) directly, `az vm extension set --resource-group tester --vm-name testa --name customScript --publisher Microsoft.Azure.Extensions --settings ./cmd.json`. The `cmd.json` can have any simple command like, 
```json
{
  "commandToExecute": "echo hello"
}
```
This should test whether or not the vm-agent is working

**Comment by koralski** [comment]
Also make sure your input script is saved with the right encoding (ANSI instead of UNICODE).
Run the CLI command with --verbose option to see the input json. I suspect the script is messed up at the input before it reaches the VM.

**Comment by koralski** [comment]
We addressed the issue and the fix will be deployed next week. The problem was that script was send in Unicode instead of UTF8. Some/most versions of Linux can handle the Unicode but some throw the error reported.

**Comment by koralski** [comment]
This fix is deployed.

**Comment by thisSIDEofRANDOM** [comment]
Thanks @koralski and @yugangw-msft !!
